### ISSUE TITLE
Remote parameterized types not behaving as expected

### ISSUE THREAD

**Issue by stocks29** [pylon/spect#15]
Parameterized types seem to be working locally, but are failing with remote types. I tried to distill this example down as best I could. 

With this code...

```elixir
defmodule MyTypes do
  @type maybe(t) :: t | nil
end

defmodule MyModule do
  @type rating :: 1 | 2 | 3 | 4 | 5

  @type t :: %__MODULE__{
    rating: MyTypes.maybe(rating())
  }
  defstruct rating: nil
end

defmodule MyModule2 do
  @typep maybe(t) :: t | nil
  @type rating :: 1 | 2 | 3 | 4 | 5

  @type t :: %__MODULE__{
    rating: maybe(rating())
  }
  defstruct rating: nil
end
```

Using `MyModule2` with the local parameterized type, things work as expected:

```elixir
iex(8)> Spect.to_spec(%{"rating" => 5}, MyModule2)
{:ok, %MyModule2{rating: 5}}
```

Using `MyModule` with the remote parameterized type, things fail:

```elixir
iex(9)> Spect.to_spec(%{"rating" => 5}, MyModule)
{:error,
 %Spect.ConvertError{
   message: "expected: union of [{:var, 2, :t}, {:atom, 0, nil}], found: 5"
 }}
```

Unrelated, but in my code I have some pretty large nested structs and it can be difficult to figure out which field is failing to decode. It turns into trial-and-error of removing fields until it passes. If it's possible, it would be great to call out in the error message specifically which field is failing at the deepest level.

**Comment by stocks29** [comment]
I did a bit more research on this. Here are my findings:

```elixir
defmodule MyTypes do
  @type maybe(t) :: t | nil
  @type rating() :: 1 | 2 | 3 | 4 | 5
end

defmodule OtherModule do
  @type our_rating() :: 1 | 2 | 3 | 4 | 5
end

defmodule MyModule do
  @type t :: %__MODULE__{
    rating: MyTypes.maybe(MyTypes.rating())
  }
  defstruct rating: nil
end

defmodule MyModule2 do
  @typep maybe(t) :: t | nil
  @type rating() :: 1 | 2 | 3 | 4 | 5

  @type t :: %__MODULE__{
    rating: maybe(rating())
  }
  defstruct rating: nil
end

defmodule MyModule3 do
  @type rating() :: :foo

  @type t :: %__MODULE__{
    rating: MyTypes.maybe(rating())
  }
  defstruct rating: nil
end

defmodule MyModule4 do
  @type my_rating() :: 1 | 2 | 3 | 4 | 5

  @type t :: %__MODULE__{
    rating: MyTypes.maybe(MyModule4.my_rating())
  }
  defstruct rating: nil
end

defmodule MyModule5 do
  @type my_rating() :: 1 | 2 | 3 | 4 | 5

  @type t :: %__MODULE__{
    rating: MyTypes.maybe(my_rating())
  }
  defstruct rating: nil
end

defmodule MyModule6 do
  @type t :: %__MODULE__{
    rating: MyTypes.maybe(OtherModule.our_rating())
  }
  defstruct rating: nil
end
```

```elixir
# works as expected
iex(4)> Spect.to_spec(%{"rating" => 5}, MyModule)
{:ok, %MyModule{rating: 5}}

# also works, but I believe it's referencing `rating()` type from `MyTypes` instead of the `MyModule2.rating()` which is a bug
iex(5)> Spect.to_spec(%{"rating" => 5}, MyModule2)
{:ok, %MyModule2{rating: 5}}

# this should fail, since the `MyModule3.rating()` should only allow `:foo`.  I believe it's referencing `MyTypes.rating()` instead which is an error
iex(6)> Spect.to_spec(%{"rating" => 5}, MyModule3)
{:ok, %MyModule3{rating: 5}}

# this should work but is likely failing because of explanation above
iex(7)> Spect.to_spec(%{"rating" => :foo}, MyModule3)
{:error,
 %Spect.ConvertError{
   message: "expected: union of [{:var, 2, :t}, {:atom, 0, nil}], found: :foo"
 }}

# I thought by using the fully qualified type name, this would work but it doesn't
iex(8)> Spect.to_spec(%{"rating" => 5}, MyModule4)
{:error,
 %Spect.ConvertError{
   message: "expected: union of [{:var, 2, :t}, {:atom, 0, nil}], found: 5"
 }}

# I believe this is failing because it's looking for `my_rating()` within `MyTypes` module.
iex(9)> Spect.to_spec(%{"rating" => 5}, MyModule5)
{:error,
 %Spect.ConvertError{
   message: "expected: union of [{:var, 2, :t}, {:atom, 0, nil}], found: 5"
 }}

# using remote types from other modules works with parameterized types
iex(10)> Spect.to_spec(%{"rating" => 5}, MyModule6)
{:ok, %MyModule6{rating: 5}}
```

In summary, it seems the problem is when using a remote parameterized type, the parameter lookup begins within the remote module instead of the caller's module. If another module is referenced by the type param, the proper type is looked up and things work as expected. However if the type param is within the caller the lookup fails.

**Comment by brentspell** [comment]
Yep, the problem here is that the source module isn't being threaded through when instantiating the parameterized type. This should be fixed in release 0.4.1 by #16.
